### PR TITLE
feat: implement basic functionality of OFREP provider

### PR DESCRIPTION
**Pull request by federicobond**

### CODE REVIEW

File: providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py
Comment (federicobond) [reply providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py:116]: Correct, I am not validating the response shape at the moment.

File: mypy.ini
Comment (beeme1mr) [inline mypy.ini:20]: Just for my own understanding, why is this needed?

File: providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py
Comment (beeme1mr) [inline providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py:40]: Where does this version come from? I wonder if it should match the library. If so, we could use Release Please to manage it for us.

**Review by beeme1mr — APPROVED** [review]

File: providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py
Comment (Kavindu-Dodan) [reply providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py:40]: You can give the option to accept a custom header provider callback, so we don't have to hardcode these values. Also, this will solve any authentication header requirements.

See Go implementation here - https://github.com/open-feature/go-sdk-contrib/blob/main/providers/ofrep/provider.go#L71-L74

File: providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py
Comment (Kavindu-Dodan) [inline providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py:95]: When will we perform the type of the returned flag ? In both JS & Go we perform a check [1] [2]

[1] - https://github.com/open-feature/js-sdk-contrib/blob/main/libs/providers/ofrep-web/src/lib/ofrep-web-provider.ts#L284-L286
[2] - https://github.com/open-feature/go-sdk-contrib/blob/main/providers/ofrep/internal/evaluate/flags.go#L38

**Comment by Kavindu-Dodan** [comment]
Good start 🙌  

My main concern is [this](https://github.com/open-feature/python-sdk-contrib/pull/88#discussion_r1597048587)

File: providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py
Comment (federicobond) [reply providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py:95]: We have a type guard [in the client](https://github.com/open-feature/python-sdk/blob/bd0bc1e2b7a28b1f1dcd50b63b61214131968925/openfeature/client.py#L440). Should we typecheck in the provider too?

File: providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py
Comment (federicobond) [reply providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py:40]: Thanks! Can you expand on why you take a callback instead of a static value? Not sure I fully understand that use case.

File: mypy.ini
Comment (gruebel) [reply mypy.ini]: fixed it 🙂

File: providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py
Comment (gruebel) [inline providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py:127]:
```suggestion
    def _handle_error(self, exception: requests.RequestException) -> NoReturn:
```
this fits better the purpose of the method, you need to import it from `typing`

File: providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py
Comment (gruebel) [inline providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py:177]:
```suggestion
    if re.match(r"^\s*[1-9][0-9]+\s*$", retry_after):
```
how about making sure, it can't be prefixed with a `0` or is `0` in general ok as value?

**Review by gruebel — COMMENTED** [review]
nice job technically, just added some small suggestions

File: providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py
Comment (Kavindu-Dodan) [reply providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py:40]: @federicobond  If we accept static value, we must deal with expired tokens. But by outsourcing it to a callback, we can expect the callback to handle refresh steps

File: providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py
Comment (Kavindu-Dodan) [reply providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py:95]: Good point. Most provider implementations I have seen perform this type validation at the provider side. IMO this is better so the provider can track its own errors, telemetry and could provide better error messages.

Comment (Kavindu-Dodan) [reply providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py:95]: Also, this is the client spec PR for ofrep where we define a type validation from the provider implementation [1]

[1] - https://github.com/open-feature/protocol/pull/14/files#diff-e32277be0442fee2ef2a5394baa0d98437fce4a32c3b1a6bd070747d4294f86cR51

File: providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py
Comment (federicobond) [reply providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py:177]: It's semantically wrong from the provider side, but it's pretty harmless as far as I can see. Should we have a special treatment for zero?

File: providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py
Comment (federicobond) [reply providers/openfeature-provider-ofrep/src/openfeature/contrib/provider/ofrep/__init__.py:53]: Understood, thank you!

**Comment by lukas-reining** [comment]
Hey @open-feature/sdk-python-maintainers, what is missing from your side, especially @federicobond to get this released?
It would be great for us to have OFREP in Python, so if I can help to get it through, I would be happy to help.

**Comment by beeme1mr** [comment]
@lukas-reining, I think it's worth getting this out there. I'll merge when the CI passes. Thanks for kicking this off, @federicobond, and sorry for the delay.

**Pull request merged**
